typing/rec_check.ml

```diff
@@ -306,338 +322,648 @@ let classify_expression : Typedtree.expression -> sd =
         Dynamic
   in classify_expression Ident.empty
 
-let rec expression : Env.env -> Typedtree.expression -> Use.t =
-  fun env exp -> match exp.exp_desc with
+let remove_pat pat env =
+  Env.remove_list (pat_bound_idents pat) env
+
+let remove_patlist pats env =
+  List.fold_right remove_pat pats env
+
+(* Usage mode judgments.
+
+   There are two main groups of judgment functions:
+
+   - Judgments of the form "Gamma |- ... : m"
+     compute the environment of a subterm from its mode,
+     so the corresponding function has type [... -> mode -> Env.t].
+
+     We write ['a -> term_judg] in this case.
+
+   - Judgments of the form "Gamma |- ... : m -| Gamma'"
+     correspond to binding constructs that have both an
+     exterior environment (the environment of the whole term)
+     and an interior environment (the environment after the binding
+     construct has introduced new names in scope).
+
+     For example, a toplevel let-binding could be given
+     the following rule:
+
+       G |- e : m + m'
+       -------------------------------
+       G |- (let x = e : m) -| x:m', G
```

Comment:
There's an issue with this rule I think (name collision on `G`; and the final output of the rule is likely not just one of the two `G`s in the scope).
Would the correct rule be something like as follows?
```
       G |- e : m + m'
       -----------------------------------
       x:m', G, G' |- (let x = e : m) -| x:m', G'
```

Reply:
(done)